balancer/frontend-v2
# feature: Use Static or VaultModel simulations for generalised joins and exits

# Description

### Use VaultModel (TS Math) and Static (onchain calls) simulations to calculate token amounts for Generalised joins/exits.

Static mode is more accurate and is used if user has:
1. Approved relayer (joins and exits)
2. Approved all tokens they are joining with (joins only)
3. Has enough token balance (joins and exits)

Otherwise, VaultModel simulation is used.

### Other changes:
- Make `useTokenApprovalActions` `tokenAddresses` parameter reactive
- join pool composable returns both relayer approval action and token approval actions (if any)

### TODO:
- Waiting for an SDK update to fix the incorrect scaling for some tokens when simulating Exit with VaultModel. (Euler pool is affected) 
  - ![Screenshot 2023-02-20 at 7 59 31](https://user-images.githubusercontent.com/28311328/220022330-0cfb6da3-08ad-41e0-9ebd-7b79b1196556.png)

## Type of change


## How should this be tested?

You can test manually with pool: http://localhost:8080/#/polygon/pool/0x48e6b98ef6329f8f0a30ebb8c7c960330d64808500000000000000000000075b/

Use https://revoke.cash/ to revoke some token approvals to see the the simulation type is correct.

**Simulation types:**
- Static (=2), Should only run when relayer and tokens in are approved (will throw an error otherwise), and if preview modal is open.
- VaultModel (=1), When the condition for the Static simulation is not met.
- Tenderly (=0), Is not used at all any longer

| Static Simulation | VaultModel Simulation |
| ----------------- | ---------------------- |
|  ![Screenshot 2023-02-13 at 18 45 33](https://user-images.githubusercontent.com/28311328/218519714-9987230c-7166-45d9-a90d-2c105560b161.png) | ![Screenshot 2023-02-13 at 18 46 32](https://user-images.githubusercontent.com/28311328/218519789-aaf35096-f34a-458a-8ec8-8688e49d8406.png) |

## Visual context

Please provide any relevant visual context for UI changes or additions. This could be static screenshots or a loom screencast.

## Checklist:
